src/libcore/hint.rs

```diff
@@ -72,3 +72,22 @@ pub fn spin_loop() {
         asm!("yield" ::: "memory" : "volatile");
     }
 }
+
+/// A function that is opaque to the optimizer, to allow benchmarks to
+/// pretend to use outputs to assist in avoiding dead-code
+/// elimination.
+///
+/// This function is a no-op, and does not even read from `dummy`.
+#[cfg_attr(any(target_arch = "asmjs", target_arch = "wasm32"), inline(never))]
```

Comment:
Instead of a duplicate `#[cfg]` listing here could there perhaps be a dummy function below in the block itself? Something like:

```rust
#[cfg(...)] {
    #[inline(never)]
    fn black_box(_: usize) {}
    black_box(&dummy as *const T as usize);
}
```

Reply:
I fixed this to do a volatile read instead, that forces LLVM to materialize the value, and is what criterion uses on its stable implementation of black box. A dummy `#[inline(never)]` function feels like a very brittle way of achieving that.

Reply:
Err, I feel that a change like this merits more discussion/consideration than a drive-by change in a PR whose title suggests no behavior change :/

I don’t have a reference at hand, but I recall some discussion that, for large values, a volatile read would have a measurable impact on benchmark numbers. Then there was a proposal of only reading the first byte, and a remark to be careful about zero-size types.